realm/realm-library/src/main/cpp/io_realm_internal_Collection.cpp

```diff
@@ -0,0 +1,482 @@
+Java_io_realm_internal_Collection_nativeGetRow(JNIEnv *env, jclass, jlong native_ptr, jint index)
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        auto wrapper = reinterpret_cast<ResultsWrapper*>(native_ptr);
+        auto row = wrapper->get_results().get(static_cast<size_t>(index));
+        return reinterpret_cast<jlong>(new Row(std::move(row)));
+    } CATCH_STD()
+    return reinterpret_cast<jlong>(nullptr);
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_Collection_nativeFirstRow(JNIEnv *env, jclass, jlong native_ptr)
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        auto wrapper = reinterpret_cast<ResultsWrapper*>(native_ptr);
+        auto optional_row = wrapper->get_results().first();
+        if (optional_row) {
+            return reinterpret_cast<jlong>(new Row(std::move(optional_row.value())));
+        }
+    } CATCH_STD()
+    return reinterpret_cast<jlong>(nullptr);
+
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_Collection_nativeLastRow(JNIEnv *env, jclass, jlong native_ptr)
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        auto wrapper = reinterpret_cast<ResultsWrapper*>(native_ptr);
+        auto optional_row = wrapper->get_results().last();
+        if (optional_row) {
+            return reinterpret_cast<jlong>(new Row(std::move(optional_row.value())));
+        }
+    } CATCH_STD()
+    return reinterpret_cast<jlong>(nullptr);
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_Collection_nativeClear(JNIEnv *env, jclass, jlong native_ptr)
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        auto wrapper = reinterpret_cast<ResultsWrapper*>(native_ptr);
+        wrapper->get_results().clear();
+        // Refresh snapshot
+        wrapper->refresh_snapshot();
+    } CATCH_STD()
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_Collection_nativeSize(JNIEnv *env, jclass, jlong native_ptr)
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        auto wrapper = reinterpret_cast<ResultsWrapper*>(native_ptr);
+        return static_cast<jlong>(wrapper->get_results().size());
+    } CATCH_STD()
+    return 0;
+}
+
+JNIEXPORT jobject JNICALL
+Java_io_realm_internal_Collection_nativeAggregate(JNIEnv *env, jclass, jlong native_ptr, jlong column_index,
+        jbyte agg_func)
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        auto wrapper = reinterpret_cast<ResultsWrapper*>(native_ptr);
+
+        size_t index = S(column_index);
+        Optional<Mixed> value;
+        switch (agg_func) {
+            case io_realm_internal_Collection_AGGREGATE_FUNCTION_MINIMUM:
+                value = wrapper->get_results().min(index);
+                break;
+            case io_realm_internal_Collection_AGGREGATE_FUNCTION_MAXIMUM:
+                value = wrapper->get_results().max(index);
+                break;
+            case io_realm_internal_Collection_AGGREGATE_FUNCTION_AVERAGE:
+                value = wrapper->get_results().average(index);
+                // TODO: Align the behavior with ObjectStore.
```

Comment:
yes, that is settled. After some bugs fixed in OS (https://github.com/realm/realm-object-store/pull/288), I think the conclusion is let binding choose if they want to align the same behavior with OS or transform the result (null->0.0) to keep current binding's behavior. For java-binding now we have no choice but convert it to `0.0` since the `average` returns `doulbe` instead of the boxed type.